File: taurus-protect-sdk-python/taurus_protect/services/tag_service.py
"""Tag service for Taurus-PROTECT SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, List, Optional

from taurus_protect.mappers.user import tag_from_dto, tags_from_dto
from taurus_protect.models.user import Tag
from taurus_protect.services._base import BaseService

if TYPE_CHECKING:
    pass  # For OpenAPI types when available


class TagService(BaseService):
    """
    Service for tag management operations.

    Provides methods to list, get, create, and delete tags.
    Tags can be applied to wallets, addresses, and other entities
    for organization and filtering.

    Example:
        >>> # List tags
        >>> tags = client.tags.list()
        >>> for tag in tags:
        ...     print(f"{tag.name}: {tag.color}")
        >>>
        >>> # Get single tag
        >>> tag = client.tags.get("tag-123")
        >>> print(f"Tag: {tag.name}")
        >>>
        >>> # Create tag
        >>> tag = client.tags.create("Important", "#FF0000")
        >>> print(f"Created tag: {tag.id}")
        >>>
        >>> # Delete tag
        >>> client.tags.delete("tag-123")
    """

    def __init__(self, api_client: Any, tags_api: Any) -> None:
        """
        Initialize tag service.

        Args:
            api_client: The OpenAPI client instance.
            tags_api: The TagsApi service from OpenAPI client.
        """
        super().__init__(api_client)
        self._tags_api = tags_api

    def get(self, tag_id: str) -> Tag:
        """
        Get a tag by ID.

        Args:
            tag_id: The tag ID to retrieve.

        Returns:
            The tag.

        Raises:
            ValueError: If tag_id is invalid.
            NotFoundError: If tag not found.
            APIError: If API request fails.
        """
        self._validate_required(tag_id, "tag_id")

        try:
            # The API doesn't have a direct get-by-id endpoint,
            # so we use the list endpoint with id filter
            resp = self._tags_api.tag_service_get_tags(
                ids=[tag_id],
                query=None,
            )

            result = getattr(resp, "result", None)
            if result is None or len(result) == 0:
                from taurus_protect.errors import NotFoundError

                raise NotFoundError(f"Tag {tag_id} not found")

            tag = tag_from_dto(result[0])
            if tag is None:
                from taurus_protect.errors import NotFoundError

                raise NotFoundError(f"Tag {tag_id} not found")

            return tag
        except Exception as e:
            from taurus_protect.errors import APIError, NotFoundError

            if isinstance(e, (APIError, NotFoundError, ValueError)):
                raise
            raise self._handle_error(e) from e

    def list(
        self,
        limit: int = 50,
        offset: int = 0,
    ) -> List[Tag]:
        """
        List tags.

        Note: The tags API does not support pagination parameters,
        so limit and offset are provided for API consistency but
        filtering is done client-side.

        Args:
            limit: Maximum number of tags to return (must be positive).
            offset: Number of tags to skip (must be non-negative).

        Returns:
            List of tags.

        Raises:
            ValueError: If limit or offset are invalid.
            APIError: If API request fails.
        """
        if limit <= 0:
            raise ValueError("limit must be positive")
        if offset < 0:
            raise ValueError("offset cannot be negative")

        try:
            resp = self._tags_api.tag_service_get_tags(
                ids=None,
                query=None,
            )

            result = getattr(resp, "result", None)
            tags = tags_from_dto(result) if result else []

            # Apply client-side pagination since API doesn't support it
            return tags[offset : offset + limit]
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            raise self._handle_error(e) from e

    def create(self, name: str, color: str) -> Tag:
        """
        Create a new tag.

        Args:
            name: The tag name/value.
            color: The tag color (hex code or color name).

        Returns:
            The created tag.

        Raises:
            ValueError: If name or color is empty.
            APIError: If API request fails.
        """
        self._validate_required(name, "name")
        self._validate_required(color, "color")

        try:
            body = {
                "value": name,
                "color": color,
            }

            resp = self._tags_api.tag_service_create_tag(body=body)

            result = getattr(resp, "result", None)
            if result is None:
                from taurus_protect.errors import APIError

                raise APIError(500, "Failed to create tag: no result returned")

            tag = tag_from_dto(result)
            if tag is None:
                from taurus_protect.errors import APIError

                raise APIError(500, "Failed to create tag: invalid response")

            return tag
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            raise self._handle_error(e) from e

    def delete(self, tag_id: str) -> None:
        """
        Delete a tag.

        This removes the tag and all its assignments from entities.

        Args:
            tag_id: The tag ID to delete.

        Raises:
            ValueError: If tag_id is empty.
            NotFoundError: If tag not found.
            APIError: If API request fails.
        """
        self._validate_required(tag_id, "tag_id")

        try:
            self._tags_api.tag_service_delete_tag(id=tag_id)
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            raise self._handle_error(e) from e

File: taurus-protect-sdk-typescript/src/services/tag-service.ts
/**
 * Tag service for Taurus-PROTECT SDK.
 *
 * Provides methods for tag management operations.
 */

import { NotFoundError, ValidationError } from '../errors';
import type { TagsApi } from '../internal/openapi/apis/TagsApi';
import { tagFromDto, tagsFromDto } from '../mappers/user';
import type { CreateTagRequest, ListTagsOptions, Tag } from '../models/user';
import { BaseService } from './base';

/**
 * Service for tag management operations.
 *
 * Provides methods to list, get, create, and delete tags.
 * Tags can be applied to wallets, addresses, and other entities
 * for organization and filtering.
 *
 * @example
 * ```typescript
 * // List every tag
 * const tags = await tagService.list();
 * for (const tag of tags) {
 *   console.log(`${tag.name}: ${tag.color}`);
 * }
 *
 * // Get single tag
 * const tag = await tagService.get('tag-123');
 * console.log(`Tag: ${tag.name}`);
 *
 * // Create tag
 * const newTag = await tagService.create({ name: 'Important', color: '#FF0000' });
 * console.log(`Created tag: ${newTag.id}`);
 *
 * // Delete tag
 * await tagService.delete('tag-123');
 * ```
 */
export class TagService extends BaseService {
  private readonly tagsApi: TagsApi;

  /**
   * Creates a new TagService instance.
   *
   * @param tagsApi - The TagsApi instance from the OpenAPI client
   */
  constructor(tagsApi: TagsApi) {
    super();
    this.tagsApi = tagsApi;
  }

  /**
   * Gets a tag by ID.
   *
   * @param tagId - The tag ID to retrieve
   * @returns The tag
   * @throws {@link ValidationError} If tagId is invalid
   * @throws {@link NotFoundError} If tag not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const tag = await tagService.get('tag-123');
   * console.log(`Tag: ${tag.name}, Color: ${tag.color}`);
   * ```
   */
  async get(tagId: string): Promise<Tag> {
    if (!tagId || tagId.trim() === '') {
      throw new ValidationError('tagId is required');
    }

    return this.execute(async () => {
      // The API doesn't have a direct get-by-id endpoint,
      // so we use the list endpoint with id filter
      const response = await this.tagsApi.tagServiceGetTags({
        ids: [tagId],
      });

      const result = response.result;

      if (!result || result.length === 0) {
        throw new NotFoundError(`Tag ${tagId} not found`);
      }

      const tag = tagFromDto(result[0]);
      if (!tag) {
        throw new NotFoundError(`Tag ${tagId} not found`);
      }

      return tag;
    });
  }

  /**
   * Lists every tag, optionally filtered by id or query.
   *
   * The tags endpoint is not paged, so the result is every tag the reply carries; there
   * is no `limit` or `offset`, and passing either is rejected by name.
   *
   * @param options - Optional `ids` and `query` filters
   * @returns Every matching tag
   * @throws {@link ValidationError} If `limit` or `offset` is passed
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * // List every tag
   * const tags = await tagService.list();
   *
   * // Search tags
   * const matching = await tagService.list({ query: 'important' });
   * ```
   */
  async list(options?: ListTagsOptions): Promise<Tag[]> {
    // A caller still paging with limit/offset would get every tag on each call and
    // could loop forever; refuse instead of ignoring them.
    const paging = options as { limit?: unknown; offset?: unknown } | undefined;
    if (paging?.limit !== undefined || paging?.offset !== undefined) {
      throw new ValidationError(
        "tags.list is not paged: it returns every tag, so limit and offset are not accepted"
      );
    }

    return this.execute(async () => {
      const response = await this.tagsApi.tagServiceGetTags({
        ids: options?.ids,
        query: options?.query,
      });
      return tagsFromDto(response.result);
    });
  }

  /**
   * Creates a new tag.
   *
   * @param request - Tag creation request
   * @returns The created tag
   * @throws {@link ValidationError} If name or color is empty
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const tag = await tagService.create({
   *   name: 'High Priority',
   *   color: '#FF0000',
   * });
   * console.log(`Created tag: ${tag.id}`);
   * ```
   */
  async create(request: CreateTagRequest): Promise<Tag> {
    if (!request.name || request.name.trim() === '') {
      throw new ValidationError('name is required');
    }
    if (!request.color || request.color.trim() === '') {
      throw new ValidationError('color is required');
    }

    return this.execute(async () => {
      const response = await this.tagsApi.tagServiceCreateTag({
        body: {
          value: request.name,
          color: request.color,
        },
      });

      const tag = tagFromDto(response.result);

      if (!tag) {
        throw new ValidationError('Failed to create tag: no result returned');
      }

      return tag;
    });
  }

  /**
   * Deletes a tag.
   *
   * This removes the tag and all its assignments from entities.
   *
   * @param tagId - The tag ID to delete
   * @throws {@link ValidationError} If tagId is empty
   * @throws {@link NotFoundError} If tag not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * await tagService.delete('tag-123');
   * console.log('Tag deleted');
   * ```
   */
  async delete(tagId: string): Promise<void> {
    if (!tagId || tagId.trim() === '') {
      throw new ValidationError('tagId is required');
    }

    return this.execute(async () => {
      await this.tagsApi.tagServiceDeleteTag({ id: tagId });
    });
  }
}
